File: app/src/main/java/com/example/macarus0/walkiewalkie/view/OwnerContactActivity.java
package com.example.macarus0.walkiewalkie.view;

import android.app.ActivityOptions;
import android.arch.lifecycle.ViewModelProviders;
import android.content.ContentResolver;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatButton;
import android.support.v7.widget.AppCompatImageButton;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.widget.EditText;
import android.widget.ImageView;

import com.example.macarus0.walkiewalkie.R;
import com.example.macarus0.walkiewalkie.data.Owner;
import com.example.macarus0.walkiewalkie.viewmodel.WalkieViewModel;
import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;

public class OwnerContactActivity extends AppCompatActivity {

    public static final String OWNER_ID = "owner_id";
    public static final int ADD_OWNER = -1;
    public static final int SELECT_OWNER_PICTURE = 201;
    public static final int SELECT_DOG = 202;

    private static final String TAG = "OwnerContactActivity";
    @BindView(R.id.owner_photo)
    ConstraintLayout ownerPhoto;
    @BindView(R.id.owner_first_name)
    EditText ownerFirstName;
    @BindView(R.id.owner_last_name)
    EditText ownerLastName;
    @BindView(R.id.owner_phone)
    EditText ownerPhone;
    @BindView(R.id.owner_email)
    EditText ownerEmail;
    @BindView(R.id.owner_address)
    EditText ownerAddress;
    @BindView(R.id.dog_card_rv)
    RecyclerView dogCardRecyclerView;
    @BindView(R.id.add_dog_button)
    AppCompatButton addDogButton;
    @BindView(R.id.contact_save_button)
    AppCompatImageButton saveButton;
    @BindView(R.id.contact_cancel_button)
    AppCompatImageButton cancelButton;
    AppCompatImageButton editOwnerPhotoButton;
    DeletablePhotoListAdapter dogsList;

    ImageView ownerImageView;
    private long mOwnerId;
    private Owner mOwner;
    private WalkieViewModel mWalkieViewModel;

    public void setOwnerId(long mOwnerId) {
        this.mOwnerId = mOwnerId;
        this.mOwner.setOwnerId(mOwnerId);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_owner);
        ButterKnife.bind(this);

        mWalkieViewModel = ViewModelProviders.of(this).get(WalkieViewModel.class);

        ownerImageView = ownerPhoto.findViewById(R.id.photo_image);
        editOwnerPhotoButton = ownerPhoto.findViewById(R.id.edit_photo_button);
        editOwnerPhotoButton.setOnClickListener(v -> selectImage());

        saveButton.setOnClickListener(v -> saveAndExit());
        cancelButton.setOnClickListener(v -> finish());

        dogsList = new DeletablePhotoListAdapter();
        dogsList.setShowPhotoLabels(true);
        dogsList.setDeleteHandler(this::removeDog);
        dogsList.setPlaceholderImage(getDrawable(R.drawable.ic_default_dog_24dp));
        dogCardRecyclerView.setAdapter(dogsList);

        LinearLayoutManager lm = new LinearLayoutManager(this);
        lm.setOrientation(LinearLayoutManager.HORIZONTAL);
        dogCardRecyclerView.setLayoutManager(lm);
        addDogButton.setText(R.string.select_dog);
        addDogButton.setOnClickListener(view -> selectDog());

        Intent intent = getIntent();
        mOwnerId = intent.getLongExtra(OWNER_ID, ADD_OWNER);
        if (mOwnerId == ADD_OWNER) {
            showAddOwnerUI();
            mOwner = new Owner();
        } else {
            mWalkieViewModel.getOwnerById(mOwnerId).observe(this,
                    this::showOwnerUI);
            mWalkieViewModel.getOwnerDogs(mOwnerId).observe(this, dogs -> dogsList.setPhotos(dogs));
        }
    }

    void showAddOwnerUI() {
        ownerFirstName.setHint(getString(R.string.add_owner_first_name_hint));
        ownerLastName.setHint(getString(R.string.add_owner_last_name_hint));
        ownerAddress.setHint(getString(R.string.add_address_hint));
        ownerEmail.setHint(getString(R.string.add_owner_email_hint));
        ownerPhone.setHint(getString(R.string.add_owner_phone_number_hint));
        ownerImageView.setImageResource(R.drawable.ic_default_owner_24dp);
    }

    void showOwnerUI(Owner owner) {
        mOwner = owner;
        Log.i(TAG, "showOwnerUI: " + owner.getDogId1());
        ownerFirstName.setText(owner.getFirstName());
        ownerLastName.setText(owner.getLastName());
        ownerAddress.setText(owner.getAddress());
        ownerEmail.setText(owner.getEmailAddress());
        ownerPhone.setText(owner.getPhoneNumber());
        Picasso.get().load(owner.getPhotoUri()).placeholder(R.drawable.ic_default_owner_24dp).into(ownerImageView);
    }

    void selectImage() {
        Intent intent = new Intent();
        intent.setType("image/*");
        intent.setAction(Intent.ACTION_GET_CONTENT);
        startActivityForResult(Intent.createChooser(intent,
                getString(R.string.select_owner_photo_activity_title)), SELECT_OWNER_PICTURE);
    }

    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            switch (requestCode) {
                case SELECT_OWNER_PICTURE:
                    Uri selectedImageUri = data.getData();
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        final int takeFlags = data.getFlags() & Intent.FLAG_GRANT_READ_URI_PERMISSION;
                        ContentResolver resolver = getContentResolver();
                        resolver.takePersistableUriPermission(selectedImageUri, takeFlags);
                    }
                    Picasso.get().load(selectedImageUri).into(ownerImageView);
                    mOwner.setPhotoUri(selectedImageUri.toString());
                    break;
                case SELECT_DOG: {
                    long dogId = data.getLongExtra(SelectDogActivity.DOG_ID, 0);
                    if (dogId == 0) break;
                    mWalkieViewModel.addOwnerToDog(mOwner.getId(), dogId);
                    break;
                }
            }
        }

    }


    void selectDog() {
        if (mOwnerId == -1) {
            mWalkieViewModel.insertOwner(mOwner).observe(this, this::setOwnerIdAndSelectDog);
        } else {
            Intent intent = new Intent(this, SelectDogActivity.class);
            intent.putExtra(SelectDogActivity.OWNER_ID, mOwnerId);
            startActivityForResult(intent, SELECT_DOG,
                    ActivityOptions.makeSceneTransitionAnimation(this).toBundle());
        }
    }

    void setOwnerIdAndSelectDog(long ownerId) {
        setOwnerId(ownerId);
        mWalkieViewModel.getOwnerDogs(mOwnerId).observe(this, dogs -> dogsList.setPhotos(dogs));
        selectDog();
    }
    void saveAndExit() {
        mOwner.setFirstName(ownerFirstName.getText().toString());
        mOwner.setLastName(ownerLastName.getText().toString());
        mOwner.setAddress(ownerAddress.getText().toString());
        mOwner.setEmailAddress(ownerEmail.getText().toString());
        mOwner.setPhoneNumber(ownerEmail.getText().toString());
        new SaveAndExit().execute();
    }

    void removeDog(long dogId) {
        mWalkieViewModel.removeOwnerFromDog(mOwner.getId(), dogId);
    }

    private class SaveAndExit extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... voids) {
            if (mOwnerId == -1) {
                mOwnerId = mWalkieViewModel.insertOwnerSync(mOwner);
                mOwner.setOwnerId(mOwnerId);
            }
            mWalkieViewModel.updateOwnerSync(mOwner);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            finish();
        }
    }


}

File: app/src/main/java/com/example/macarus0/walkiewalkie/view/OwnerContactActivity.kt
package com.example.macarus0.walkiewalkie.view

import android.app.Activity
import android.app.ActivityOptions
import android.app.Dialog
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import butterknife.ButterKnife
import com.example.macarus0.walkiewalkie.R
import com.example.macarus0.walkiewalkie.data.Dog
import com.example.macarus0.walkiewalkie.data.Owner
import com.example.macarus0.walkiewalkie.viewmodel.WalkieViewModel
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_edit_dog.*
import kotlinx.android.synthetic.main.activity_edit_owner.*
import kotlinx.android.synthetic.main.contact_edit_header.*
import kotlinx.android.synthetic.main.photo_select.*
import kotlinx.android.synthetic.main.photo_select.view.*

class OwnerContactActivity : AppCompatActivity() {

    private lateinit var dogsList: DeletablePhotoListAdapter
    private var mOwnerId: Long = 0L
    private lateinit var mOwner: Owner
    private lateinit var mWalkieViewModel: WalkieViewModel

    private fun setOwnerId(mOwnerId: Long) {
        this.mOwnerId = mOwnerId
        this.mOwner.ownerId = mOwnerId
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_owner)
        ButterKnife.bind(this)
        mWalkieViewModel = ViewModelProviders.of(this).get(WalkieViewModel::class.java)


        editPhotoButton.setOnClickListener { _ -> selectImage() }
        contactSaveButton.setOnClickListener { _ -> saveAndExit() }
        contactCancelButton.setOnClickListener { _ -> finish() }
        contactDeleteButton.setOnClickListener{ _ -> showDeleteConfirm() }

        dogsList = DeletablePhotoListAdapter()
        dogsList.showPhotoLabels = true
        dogsList.deleteHandler = {id ->  this.removeDog(id) }
        dogsList.placeholderImage = requireNotNull(getDrawable(R.drawable.ic_default_dog_24dp))
        dogCardRecyclerView.adapter = dogsList

        val lm = LinearLayoutManager(this)
        lm.orientation = LinearLayoutManager.HORIZONTAL
        dogCardRecyclerView.layoutManager = lm
        addDogButton.setText(R.string.select_dog)
        addDogButton.setOnClickListener { _ -> selectDog() }

        val intent = intent
        mOwnerId = intent.getLongExtra(OWNER_ID, ADD_OWNER.toLong())
        if (mOwnerId == ADD_OWNER.toLong()) {
            showAddOwnerUI()
            mOwner = Owner()
        } else {
            val ownerLiveData = mWalkieViewModel.getOwnerById(mOwnerId)
            ownerLiveData.observe(this, Observer {
                showOwnerUI(it)
                ownerLiveData.removeObservers(this)
            })
            mWalkieViewModel.getOwnerDogs(mOwnerId).observe(this, Observer<List<Dog>> { dogsList.photos = it })
        }
    }

    private fun showAddOwnerUI() {
        ownerFirstName.hint = getString(R.string.add_owner_first_name_hint)
        ownerLastName.hint = getString(R.string.add_owner_last_name_hint)
        ownerAddress.hint = getString(R.string.add_address_hint)
        ownerEmail.hint = getString(R.string.add_owner_email_hint)
        ownerPhone.hint = getString(R.string.add_owner_phone_number_hint)
        ownerPhoto.photoSelectImageView.setImageResource(R.drawable.ic_default_owner_24dp)
    }

    private fun showOwnerUI(owner: Owner) {
        mOwner = owner
        ownerFirstName!!.setText(owner.firstName)
        ownerLastName!!.setText(owner.lastName)
        ownerAddress!!.setText(owner.address)
        ownerEmail!!.setText(owner.emailAddress)
        ownerPhone!!.setText(owner.phoneNumber)
        Picasso.get().load(owner.photoUri).placeholder(R.drawable.ic_default_owner_24dp).into(ownerPhoto.photoSelectImageView)
    }

    private fun selectImage() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
        intent.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION)
        intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intent.type = "image/*"
        startActivityForResult(Intent.createChooser(intent,
                getString(R.string.select_owner_photo_activity_title)), SELECT_OWNER_PICTURE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            when (requestCode) {
                SELECT_OWNER_PICTURE -> {
                    val selectedImageUri = data!!.data
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        val takeFlags = data.flags and Intent.FLAG_GRANT_READ_URI_PERMISSION
                        val resolver = contentResolver
                        resolver.takePersistableUriPermission(selectedImageUri!!, takeFlags)
                    }
                    Picasso.get().load(selectedImageUri).into(ownerPhoto.photoSelectImageView)
                    mOwner.photoUri = selectedImageUri!!.toString()
                }
                SELECT_DOG -> {
                    val dogId = data!!.getLongExtra(SelectDogActivity.DOG_ID, 0)
                    if (dogId == 0L) return
                    mWalkieViewModel.addOwnerToDog(mOwner.id, dogId)
                }
            }
        }

    }


    private fun selectDog() {
        if (mOwnerId == -1L) {
            mWalkieViewModel.insertOwner(mOwner).observe(this, Observer<Long> { this.setOwnerIdAndSelectDog(requireNotNull(it)) })
        } else {
            val intent = Intent(this, SelectDogActivity::class.java)
            intent.putExtra(SelectDogActivity.OWNER_ID, mOwnerId)
            startActivityForResult(intent, SELECT_DOG,
                    ActivityOptions.makeSceneTransitionAnimation(this).toBundle())
        }
    }

    private fun setOwnerIdAndSelectDog(ownerId: Long) {
        setOwnerId(ownerId)
        mWalkieViewModel.getOwnerDogs(mOwnerId).observe(this, Observer<List<Dog>> { dogsList.photos = it })
        selectDog()
    }

    private fun saveAndExit() {
        mOwner.firstName = ownerFirstName!!.text.toString()
        mOwner.lastName = ownerLastName!!.text.toString()
        mOwner.address = ownerAddress!!.text.toString()
        mOwner.emailAddress = ownerEmail!!.text.toString()
        mOwner.phoneNumber = ownerEmail!!.text.toString()
        SaveAndExit().execute()
    }

    fun removeDog(dogId: Long) {
        mWalkieViewModel.removeOwnerFromDog(mOwner.id, dogId)
    }

    private fun showDeleteConfirm() {
        val popupWindow = Dialog(this)
        popupWindow.setContentView(R.layout.popup_window)

        var textView = popupWindow.findViewById<TextView>(R.id.popupMessageTextView)
        textView.setText(R.string.confirm_delete_owner)
        var okButton = popupWindow.findViewById<Button>(R.id.popupOk)
        var cancelButton = popupWindow.findViewById<Button>(R.id.popupCancel)

        okButton.setText(R.string.ok)
        okButton.setOnClickListener{deleteOwner()}
        cancelButton.setText(R.string.cancel)
        cancelButton.setOnClickListener{popupWindow.dismiss()}
        popupWindow.show()
    }

    private fun deleteOwner() {
        Toast.makeText(this, R.string.owner_deleted, Toast.LENGTH_SHORT).show()
        DeleteAndExit().execute()
    }

    private inner class DeleteAndExit : AsyncTask<Void, Void, Void>() {

        override fun doInBackground(vararg voids: Void): Void? {
            if (mOwnerId != -1L) {
                mWalkieViewModel.deleteOwner(mOwnerId)
            }
            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            finish()
        }
    }


    private inner class SaveAndExit : AsyncTask<Void, Void, Void>() {

        override fun doInBackground(vararg voids: Void): Void? {
            if (mOwnerId == -1L) {
                mOwnerId = mWalkieViewModel.insertOwnerSync(mOwner)
                mOwner.ownerId = mOwnerId
            }
            mWalkieViewModel.updateOwnerSync(mOwner)
            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            finish()
        }
    }

    companion object {

        const val OWNER_ID = "owner_id"
        const val ADD_OWNER = -1
        const val SELECT_OWNER_PICTURE = 201
        const val SELECT_DOG = 202

    }


}
